internal/privatestate/data.go

```diff
@@ -0,0 +1,200 @@
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// Data contains private state data for the framework and providers.
+type Data struct {
+	// Potential future usage:
+	// Framework contains private state data for framework usage.
+	Framework map[string][]byte
+
+	// Provider contains private state data for provider usage.
+	Provider ProviderData
+}
+
+// Bytes returns a JSON encoded slice of bytes containing the merged
+// framework and provider private state data.
+func (d *Data) Bytes(_ context.Context) ([]byte, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if d == nil || (len(d.Framework)+len(d.Provider)) == 0 {
+		return nil, nil
+	}
+
+	mergedMap := make(map[string][]byte, len(d.Framework)+len(d.Provider))
+
+	for k, v := range d.Framework {
+		mergedMap[k] = v
+	}
+
+	for k, v := range d.Provider {
+		mergedMap[k] = v
+	}
+
+	bytes, err := json.Marshal(mergedMap)
+	if err != nil {
+		diags.AddError(
+			"Error Encoding Private State",
+			fmt.Sprintf("An error was encountered when encoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	return bytes, diags
+}
+
+// NewData creates a new Data based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewData(ctx context.Context, data []byte) (*Data, diag.Diagnostics) {
+	var (
+		dataMap map[string][]byte
+		diags   diag.Diagnostics
+	)
+
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		diags.AddError(
+			"Error Decoding Private State",
+			fmt.Sprintf("An error was encountered when decoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	output := Data{
+		Framework: make(map[string][]byte),
+		Provider:  make(map[string][]byte),
+	}
+
+	for k, v := range dataMap {
+		if isInvalidProviderDataKey(ctx, k) {
+			output.Framework[k] = v
+			continue
+		}
+
+		output.Provider[k] = v
+	}
+
+	return &output, diags
+}
+
+// ProviderData contains private state data for provider usage.
+type ProviderData map[string][]byte
+
+// GetKey returns the private state data associated with the given key.
+//
+// If the key is reserved for framework usage, an error diagnostic
+// is returned. If the key is valid, but private state data is not found,
+// nil is returned.
+//
+// The naming of keys only matters in context of a single resource,
+// however care should be taken that any historical keys are not reused
+// without accounting for older resource instances that may still have
+// older data at the key.
+func (d *ProviderData) GetKey(ctx context.Context, key string) ([]byte, diag.Diagnostics) {
```

Comment:
Does `GetKey` need to be declared on a pointer receiver given that `ProviderData` is a `map[string][]byte`?

Reply:
It's probably fine either which way -- read-only methods shouldn't need pointer receiver handling unless handling a `nil` pointer of that type is important, such as panic prevention.